core/node/proof_data_handler/src/tee_request_processor.rs

```diff
@@ -0,0 +1,134 @@
+
+pub type TeeProofGenerationDataResponse = GenericProofGenerationDataResponse<TeeVerifierInput>;
+
+#[derive(Clone)]
+pub(crate) struct TeeRequestProcessor {
+    blob_store: Arc<dyn ObjectStore>,
+    pool: ConnectionPool<Core>,
+    config: ProofDataHandlerConfig,
+}
+
+impl TeeRequestProcessor {
+    pub(crate) fn new(
+        blob_store: Arc<dyn ObjectStore>,
+        pool: ConnectionPool<Core>,
+        config: ProofDataHandlerConfig,
+    ) -> Self {
+        Self {
+            blob_store,
+            pool,
+            config,
+        }
+    }
+
+    pub(crate) async fn get_proof_generation_data(
+        &self,
+        request: Json<TeeProofGenerationDataRequest>,
+    ) -> Result<Json<TeeProofGenerationDataResponse>, RequestProcessorError> {
+        tracing::info!("Received request for proof generation data: {:?}", request);
+
+        let mut connection = self
+            .pool
+            .connection()
+            .await
+            .map_err(|_| RequestProcessorError::Sqlx(SqlxError::PoolClosed))?;
+
+        let l1_batch_number_result = connection
+            .tee_proof_generation_dal()
+            .get_next_block_to_be_proven(self.config.proof_generation_timeout())
+            .await;
+        let l1_batch_number = match l1_batch_number_result {
+            Some(number) => number,
+            None => return Ok(Json(TeeProofGenerationDataResponse::Success(None))),
+        };
+
+        let tee_verifier_input: TeeVerifierInput = self
+            .blob_store
+            .get(l1_batch_number)
+            .await
+            .map_err(RequestProcessorError::ObjectStore)?;
+
+        Ok(Json(TeeProofGenerationDataResponse::Success(Some(
+            Box::new(tee_verifier_input),
+        ))))
+    }
+
+    pub(crate) async fn submit_proof(
+        &self,
+        Path(l1_batch_number): Path<u32>,
+        Json(payload): Json<SubmitTeeProofRequest>,
+    ) -> Result<Json<SubmitProofResponse>, RequestProcessorError> {
+        let l1_batch_number = L1BatchNumber(l1_batch_number);
+        let mut connection = self
+            .pool
+            .connection()
+            .await
+            .map_err(|_| RequestProcessorError::Sqlx(SqlxError::PoolClosed))?;
+        let mut dal = connection.tee_proof_generation_dal();
+
+        match payload {
+            SubmitTeeProofRequest::Proof(proof) => {
+                tracing::info!(
+                    "Received proof {:?} for block number: {:?}",
+                    proof,
+                    l1_batch_number
+                );
+                dal.save_proof_artifacts_metadata(
+                    l1_batch_number,
+                    &proof.signature,
+                    &proof.pubkey,
+                    &proof.proof,
+                    TeeType::Sgx,
+                )
+                .await
+                .map_err(RequestProcessorError::Sqlx)?;
+            }
+            SubmitTeeProofRequest::SkippedProofGeneration => {
```

Comment:
Is this logic actually used? I was thinking we don't even create jobs when TEEs are not enabled.

Reply:
@RomanBrodetski I'm not sure if I understand, can you pls elaborate on that? I tried to grasp [the reasoning](https://matter-labs-workspace.slack.com/archives/C05ANUCGCKV/p1694763650123479) behind using `SkippedProofGeneration` for rate limiting and thought we could apply the same logic to TEEs if we want to rate limit for any reason. I'm not sure why you're linking it to the TEE enable/disable toggle switch. 🤔

Reply:
We can skip proof generation on testnet/local envs. Locally it makes sense as we don't run a prover. On testnet we do run a prover, but proving is expensive, so we apply sampling there -- we only generate _some_ proves, not for every batch.
With TEE, ability to skip proofs doesn't seem to make sense -- since we may simply work without it. If we don't need logic for skipping jobs, maybe it makes sense to remove this functionality completely?

Reply:
[Fixed](https://github.com/matter-labs/zksync-era/pull/1993/commits/eb145f4ec07c9404d97321d4e304e3da45e5d886). ✅